### PR TITLE
/nodes POST and /nodes/<node_id> DELETE enabled with optimizations

### PR DESCRIPTION
**Pull request by zzxgzgz**
/nodes POST and /nodes/<node_id> DELETE enabled.

Optimizations:

1. Put util functions into utils.py.
2. Get zgc CIDR range from configmap, and it also matches the zgc network created for KIND in create_cluster.sh.
3. Only updates modified Droplets, instead of all of them.
4. Added logic for errors. 
5. Added logic to distribute IPs in ascending order, and give up IPs in descending order. 
6. Other small fixes.

### CODE REVIEW

File: test/yaml/droplet.yaml [hunk 1]
```diff
@@ -8,7 +8,11 @@ kind: Droplet
...
-  ip: "10.0.0.1"
+  mac: 
+  - "0000:0000:0000:0001"
+  - "0000:0000:0000:0002"
+  ip: 
+  - "10.0.0.1"
+  - "10.0.0.2"
   interface: "eth0"
-  status: "Init"
+  status: "Init"
```
Comment (liangbin-pub) [inline test/yaml/droplet.yaml]: Add newline at end of file

File: src/mgmt/services/manager/project/api/vpcs.py [hunk 2]
```diff
@@ -17,33 +17,12 @@
...
-        gw["mac"] = int_to_mac(ipint | MAC_PRIVATE)
-        gws.append(gw.copy())
-    return gws
-
-def extendVpcResp(vpc):
-    respond = vpc.to_json()
-    respond['port_ibo'] = settings.activeZgc["port_ibo"]
-    respond['gws'] = getGWsFromIpRange(settings.activeZgc["ip_start"], settings.activeZgc["ip_end"])
-    return respond
+
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/vpcs.py]: Don't make extendVpcResp a utility as it's specific to VPC, make  getGWsFromIpRange() an utility instead

File: src/mgmt/services/manager/project/api/vpcs.py [hunk 3]
```diff
@@ -90,4 +69,4 @@ def single_vpc(vpc_id):
 
 
 if __name__ == '__main__':
-    app.run()
+    app.run()
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/vpcs.py]: no whitespace changes, add newline at end of file

File: src/mgmt/services/manager/project/api/utils.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+from project.api import settings
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/utils.py]: Add license header

File: src/mgmt/services/manager/project/api/utils.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+MAC_PRIVATE = 0x828300000000
+
+# Convert to GW array
+def int_to_mac(macint):
+    return ':'.join(['{}{}'.format(a, b)
+                     for a, b
+                     in zip(*[iter('{:012x}'.format(macint))]*2)])
+
+def getGWsFromIpRange(firstIp,lastIp):
+    start= reduce(lambda a,b: a<<8 | b, map(int, firstIp.split(".")))
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/utils.py]: make the reduce() an utility as it's been used 3 times here

File: src/mgmt/services/manager/project/api/utils.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    gws = []
+    for ipint in range(start,end+1):
+        gw = {}
+        gw["ip"] = ".".join(map(lambda n: str(ipint>>n & 0xFF), [24,16,8,0]))
+        gw["mac"] = int_to_mac(ipint | MAC_PRIVATE)
+        gws.append(gw.copy())
+    return gws
+
+def extendVpcResp(vpc):
+    respond = vpc.to_json()
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/utils.py]: Don't put this func here as that can lead to import loop and it's not utility

File: src/mgmt/services/manager/project/api/utils.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    respond['port_ibo'] = settings.activeZgc["port_ibo"]
+    respond['gws'] = getGWsFromIpRange(settings.activeZgc["ip_start"], settings.activeZgc["ip_end"])
+    return respond
+
+
+def get_mac_from_ip(ip_string):
+    start= reduce(lambda a,b: a<<8 | b, map(int, ip_string.split('.')))
+    mac = int_to_mac(start | MAC_PRIVATE)
+    return mac
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/utils.py]: Check missing end of file newline

File: src/mgmt/services/manager/project/api/settings.py [hunk 8]
```diff
@@ -3,7 +3,24 @@
...
     activeZgc = {}
+    global zgc_cidr_range
+    # Give it a default value.
+    zgc_cidr_range = '10.0.0.0/16'
+    try:
+        api_response = core_api.list_config_map_for_all_namespaces(label_selector='name=zgc-cluster-config')
+        logger.info('Result for getting configmap: {}'.format(api_response))
+        zgc_cidr_range = api_response.items[0].data['cidr']
+    except ApiException as e:
+        logger.error('Exception when updating exsiting droplet: {}'.format(e))
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/settings.py]: The error message is not correctly indicate when it happened, also add newline at end of file.
Add MAC_PRIVATE = 0x828300000000 to config map as well

File: src/mgmt/services/manager/project/api/nodes.py [hunk 9]
```diff
@@ -62,11 +227,58 @@ def single_node(node_id):
...
+
         db.session.delete(node)
         db.session.commit()
         response_object = {}
     return jsonify(response_object)
 
 
 if __name__ == '__main__':
-    app.run()
+    app.run()
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py]: why there are these kind of whitespace changes? don't remove newline from end of file

File: src/mgmt/services/manager/project/api/nodes.py [hunk 10]
```diff
@@ -9,31 +9,196 @@
...
+        logger.info('Response for create droplet: {}'.format(create_response))
+    except ApiException as e:
+        logger.error('Exception when Creating droplets: {}'.format(e))
+
 @nodes_blueprint.route('/nodes', methods=['GET', 'POST'])
 def all_nodes():
     if request.method == 'POST':
         post_data = request.get_json()
         post_data['node_id'] = str(uuid.uuid4())
+        response_object = deepcopy(post_data)
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py]: Don't see any reason to make this change

File: src/mgmt/services/manager/project/api/nodes.py [hunk 11]
```diff
@@ -9,31 +9,196 @@
...
+from kubernetes.client.rest import ApiException
+from kubernetes import client, config
 from project.api.models import Node
 from project import db
+from project.api.models import Zgc
+from project.api.utils import getGWsFromIpRange, get_mac_from_ip
+import json
+import operator
+from project.api.settings import zgc_cidr_range
+from copy import deepcopy
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:25]: See comment below, shouldn't need it

File: src/mgmt/services/manager/project/api/nodes.py [hunk 12]
```diff
@@ -9,31 +9,196 @@
...
+                                                                 version='v1',
+                                                                 plural='droplets',
+                                                                 label_selector='zgc_id='+post_data['zgc_id']+',network=tenant'
+                                                                )['items']
+        zgc_ip_list = post_data['ip_control'].split('.')
+        
+        ip_range = zgc_cidr_range.split('/')
+        cidr_list = ip_range[0].split('.')
+
+        for i in range(int(ip_range[1])//8):
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:105]: Add a comment here to indicate only zgc_cidr prefix 8, 16 or 24 are supported

File: src/mgmt/services/manager/project/api/nodes.py [hunk 13]
```diff
@@ -9,31 +9,196 @@
...
+                               itf=post_data['inf_zgc'], 
+                               network='zgc', 
+                               zgc_id=post_data['zgc_id'])
+            else:
+                logger.error("There's no zgc with zgc_id: {} in the database!".format(post_data['zgc_id']))
+                return jsonify({'error':'No such zgc'})
+        # commit change to data at last
         db.session.add(Node(**post_data))
         db.session.commit()
-        response_object = post_data
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:31]: don't make this change, see comment above

File: deploy/etc/deployments/zgc_cluster_config.yaml [hunk 14]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: zgc-cluster-config
+  labels:
+    name: zgc-cluster-config
+data:
+  cidr: 10.0.0.0/24
```
Comment (liangbin-pub) [inline deploy/etc/deployments/zgc_cluster_config.yaml]: Add newline at end of file
add MAC_PRIVATE here as well

File: test/yaml/droplet.yaml (hunk 1 above)
Comment (zzxgzgz) [reply test/yaml/droplet.yaml]: Fixed

File: src/mgmt/services/manager/project/api/vpcs.py (hunk 2 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/vpcs.py]: fixed

File: src/mgmt/services/manager/project/api/vpcs.py (hunk 3 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/vpcs.py]: fixed

File: src/mgmt/services/manager/project/api/utils.py (hunk 4 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/utils.py]: fixed

File: src/mgmt/services/manager/project/api/utils.py (hunk 7 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/utils.py]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 9 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 10 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 11 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py:25]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 12 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py:105]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 13 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py:31]: fixed

File: deploy/etc/deployments/zgc_cluster_config.yaml (hunk 14 above)
Comment (zzxgzgz) [reply deploy/etc/deployments/zgc_cluster_config.yaml]: fixed

File: src/mgmt/services/manager/project/api/utils.py (hunk 5 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/utils.py]: fixed

File: src/mgmt/services/manager/project/api/utils.py (hunk 6 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/utils.py]: fxied

File: src/mgmt/services/manager/project/api/settings.py (hunk 8 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/settings.py]: fixed

File: src/mgmt/operator/obj/droplet.py [hunk 15]
```diff
@@ -36,9 +36,11 @@ def __init__(self, name, obj_api, opr_store, spec=None):
...
         self.kind = "Droplet"
         self.plural = "droplets"
-        self.ip = ""
-        self.mac = ""
-        self.phy_itf = 'eth0'
+        self.ip = []
+        self.mac = []
+        self.zgc_id = ''
+        self.phy_itf = ''
+        self.name = name
```
Comment (phudtran) [inline src/mgmt/operator/obj/droplet.py]: We can remove this self.name assignment as the call to super already takes care of it

File: src/mgmt/operator/obj/droplet.py [hunk 16]
```diff
@@ -48,10 +50,13 @@ def rpc(self):
 
     def get_obj_spec(self):
         self.obj = {
+            "name": self.name,
```
Comment (phudtran) [inline src/mgmt/operator/obj/droplet.py]: We don't need this as the name is in the metadata.

File: src/mgmt/operator/workflows/droplets/triggers.py [hunk 17]
```diff
@@ -58,7 +58,7 @@ def droplet_opr_on_droplet_provisioned(body, spec, **kwargs):
 @kopf.on.delete(group, version, RESOURCES.droplets, retries=OBJ_DEFAULTS.kopf_max_retries)
 def droplet_opr_on_droplet_delete(body, spec, **kwargs):
     param = HandlerParam()
-    param.name = kwargs['name']
+    param.name = spec['name']
```
Comment (phudtran) [inline src/mgmt/operator/workflows/droplets/triggers.py]: Not sure why this needs to be changed, but if we change it here can we change it everywhere else to be consistent?

File: src/mgmt/services/manager/project/api/settings.py [hunk 18]
```diff
@@ -3,7 +3,28 @@
...
+    global MAC_PRIVATE
+    # Give it a default 
+    MAC_PRIVATE = 0x928300000000
+    try:
+        api_response = core_api.list_config_map_for_all_namespaces(label_selector='name=zgc-cluster-config')
+        logger.info('Result for getting configmap: {}'.format(api_response))
+        zgc_cidr_range = api_response.items[0].data['cidr']
+        MAC_PRIVATE = int(api_response.items[0].data['MAC_PRIVATE'],0)
+    except ApiException as e:
+        logger.error('Exception when trying to get zgc-cluster-config: {}'.format(e))
```
Comment (phudtran) [inline src/mgmt/services/manager/project/api/settings.py]: New line

File: test/yaml/droplet.yaml [hunk 19]
```diff
@@ -8,7 +8,12 @@ kind: Droplet
...
-  ip: "10.0.0.1"
+  mac: 
+  - "0000:0000:0000:0001"
+  - "0000:0000:0000:0002"
+  ip: 
+  - "10.0.0.1"
+  - "10.0.0.2"
   interface: "eth0"
   status: "Init"
+
```
Comment (phudtran) [inline test/yaml/droplet.yaml]: new line

File: src/mgmt/services/manager/project/api/nodes.py [hunk 20]
```diff
@@ -9,31 +9,196 @@
...
+    try:
+        update_response = obj_api.replace_namespaced_custom_object(group='zeta.com',
+                                                                   version='v1',
+                                                                   plural='droplets',
+                                                                   namespace='default', 
+                                                                   name=droplet['metadata']['name'],
+                                                                   body=droplet)
+        logger.info('Response for update droplet: {}'.format(update_response))
+    except ApiException as e:
+        logger.error('Exception when updating exsiting droplet: {}'.format(e))
```
Comment (phudtran) [inline src/mgmt/services/manager/project/api/nodes.py]: exsiting  -> existing

File: src/mgmt/services/manager/project/api/nodes.py [hunk 21]
```diff
@@ -62,6 +227,53 @@ def single_node(node_id):
...
+            logger.error('Cannot find that droplet with name: {}, thus not deleting anything'.format(tenant_droplet_name))
+            return jsonify({'error': 'Droplet not found'})
+        else:
+            ip_to_assign = droplet_to_remove['spec']['ip']
+
+            ip_amount = len(ip_to_assign)
+
+            number_of_droplets_to_assign = len(all_droplets_in_zgc)
+
+            # Sort these droplets in accending order, so the droplet with the least IPs will be in the front.
```
Comment (phudtran) [inline src/mgmt/services/manager/project/api/nodes.py]: ascending -> ascending

File: src/mgmt/services/manager/project/api/nodes.py [hunk 22]
```diff
@@ -9,31 +9,196 @@
...
+                droplet_ip_list = droplet_spec['ip']
+                droplet_mac_list = droplet_spec['mac']
+                total_ip = total_ip + len(droplet_ip_list)
+                if len(droplet_ip_list) > 1 and len(droplet_mac_list) > 1:
+                    droplet_that_can_give_ip_mac.append(droplet)
+            number_of_droplets = len(all_droplets_in_zgc)
+            number_of_ip_new_droplet_gets = total_ip // (number_of_droplets + 1) 
+            ip_for_new_droplet = []
+            modified_droplets = dict()
+            if number_of_ip_new_droplet_gets > 0 : # each droplet should have 1 or more ip, assign ip / mac from exiting droplets
```
Comment (phudtran) [inline src/mgmt/services/manager/project/api/nodes.py]: exiting -> existing

File: src/mgmt/services/manager/project/api/nodes.py [hunk 23]
```diff
@@ -9,31 +9,196 @@
...
+                current_length_of_ip_list = len(ip_for_new_droplet)
+                # sort these droplets in decending order, so droplet with the most IPs will be in the front.
```
Comment (phudtran) [inline src/mgmt/services/manager/project/api/nodes.py]: decending -> descending

File: deploy/etc/crds/droplets.crd.yaml [hunk 24]
```diff
@@ -68,3 +59,43 @@ spec:
...
+          properties:
+            status:
+              type: string
+            mac:
+              type: array
+              items: string
+            ip:
+              type: array
+              items: string
+            interface:
```
Comment (phudtran) [inline deploy/etc/crds/droplets.crd.yaml:92]: I don't think we need this one as we already have a spec.itf defined above.

**Review by phudtran — CHANGES_REQUESTED** [review]
Overall looks ok to me, just some minor stuff.

File: src/mgmt/operator/obj/droplet.py (hunk 15 above)
Comment (zzxgzgz) [reply src/mgmt/operator/obj/droplet.py]: fixed

File: src/mgmt/operator/obj/droplet.py (hunk 16 above)
Comment (zzxgzgz) [reply src/mgmt/operator/obj/droplet.py]: fixed

File: src/mgmt/operator/workflows/droplets/triggers.py (hunk 17 above)
Comment (zzxgzgz) [reply src/mgmt/operator/workflows/droplets/triggers.py]: fixed

File: src/mgmt/services/manager/project/api/settings.py (hunk 18 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/settings.py]: fixed

File: test/yaml/droplet.yaml (hunk 19 above)
Comment (zzxgzgz) [reply test/yaml/droplet.yaml]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 20 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 21 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 22 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py]: fixed

File: src/mgmt/services/manager/project/api/nodes.py (hunk 23 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py]: fixed

File: deploy/etc/crds/droplets.crd.yaml (hunk 24 above)
Comment (zzxgzgz) [reply deploy/etc/crds/droplets.crd.yaml:92]: fixed

**Review by phudtran — APPROVED** [review]

**Review by liangbin-pub — APPROVED** [review]

**Pull request merged**